Code/Framework/AzCore/AzCore/Settings/SettingsRegistryImpl.h

```diff
@@ -159,5 +159,11 @@ namespace AZ
         //! Stack tracking the files currently being merged
         //! This is protected by m_settingsMutex
         AZStd::stack<AZ::IO::FixedMaxPath> m_mergeFilePathStack;
+
+        // if this is nonzero, we are in a visit operation.  It can be used to detect illegal modifications
+        // of the tree during visit.
+        mutable int m_visitDepth = 0; // mutable due to it being a debugging value used in const.
+
+        void CheckVisitMutate() const;
```

Comment:
I suggest changing the interface to be `AZStd::scoped_lock AcquireLock()` or something and performing the assert and lock acquisition in one go. (may need a unique_lock I can't remember the exact API)

Reply:
I'll take look into doing this.  I was trying to shuffle as little actual API around, but yeah, that might be safer!

Reply:
looks like it works just with scoped_lock as the return.  Even in debug RVO ends up making it so that the outside scope owns the lock

Reply:
Note that not all instances of locking this mutex require us to check for visit mutation, specifically when actually visiting, or reading data without mutating (all those const functions).  So it'd have to be named something like AquireLockAndCheckForInvalidWriteAccessDuringVisit and it wouldn't be hte only place you'd aquire the lock.

Reply:
I think what I'll do is make 2 symmetric functions, LockForReading / LockForWriting
And then just do the check in LockForWriting.  replace all instances where m_settingMutex is used directly with one of those.   Each operation is either mutating the registry in some way (including installing and removing callbacks) or is not.

Reply:
That sounds good to me.
The immutable operations of `Get/GetObject` and `Visit` so they should use LockForReading.

Reply:
Hmmm, just thinking out loud - if we're already dividing out reading vs writing locks, I wonder if there's any benefit at some point to switching this over to a shared_mutex...